modules/space/horizonsfile.cpp

```diff
@@ -0,0 +1,877 @@
+                                       const std::string& observer,
+                                       const std::string& startTime,
+                                       const std::string& stopTime,
+                                       const std::string& stepSize,
+                                       const std::string& unit)
+{
+    // Construct url for request
+    std::string url = "";
+    switch (type) {
+        case Type::Vector:
+            url = VectorUrl;
+            break;
+        case Type::Observer:
+            url = ObserverUrl;
+            break;
+        default:
+            throw ghoul::MissingCaseException();
+    }
+
+    url += fmt::format(
+        "{}'{}'",
+        Command,
+        urlEncode(target)
+    );
+
+    url += fmt::format(
+        "{}'{}'",
+        Center,
+        urlEncode(observer)
+    );
+
+    url += fmt::format(
+        "{}'{}'",
+        StartTime,
+        urlEncode(startTime)
+    );
+
+    url += fmt::format(
+        "{}'{}'",
+        StopTime,
+        urlEncode(stopTime)
+    );
+
+    if (unit.empty()) {
+        url += fmt::format(
+            "{}'{}'",
+            StepSize,
+            urlEncode(stepSize)
+        );
+    }
+    else {
+        url += fmt::format(
+            "{}'{}{}{}'",
+            StepSize,
+            urlEncode(stepSize),
+            WhiteSpace,
+            unit
+        );
+    }
+
+    return url;
+}
+
+HorizonsFile::ResultCode HorizonsFile::isValidAnswer(const json& answer) {
+    // Signature, source and version
+    if (auto signature = answer.find("signature"); signature != answer.end()) {
+
+        if (auto source = signature->find("source"); source != signature->end()) {
+            if (*source != static_cast<std::string>(ApiSource)) {
+                LWARNING(fmt::format("Horizons answer from unkown source '{}'",
+                    *source)
+                );
+            }
+        }
+        else {
+            LWARNING("Could not find source information, source might not be acceptable");
+        }
+
+        if (auto version = signature->find("version"); version != signature->end()) {
+            if (*version != static_cast<std::string>(CurrentVersion)) {
+                LWARNING(fmt::format("Unknown Horizons version '{}' found. The "
+                    "currently supported version is {}", *version, CurrentVersion));
+            }
+        }
+        else {
+            LWARNING(
+                "Could not find version information, version might not be supported"
+            );
+        }
+    }
+    else {
+        LWARNING("Could not find signature information");
+    }
+
+    // Errors
+    if (auto it = answer.find("error"); it != answer.end()) {
+        // There was an error
+        std::string errorMsg = *it;
+
+        // Projected output length (~X) exceeds 90024 line max -- change step-size
+        if (errorMsg.find("Projected output length") != std::string::npos) {
```

Comment:
Do we maybe need a small helper function for doing this check? is done a LOT :)

Reply:
Can we use C++23 features yet? I guess not? Will be nice when we have the `std::contains` function https://en.cppreference.com/w/cpp/string/basic_string/contains

Reply:
Marked as CPP23 future improvements